File: vcccpp/src/vcc.cpp
/*

vccycleport.cpp
 - Input:
     * vector<umComponent> Components ={}
     * vector<tupConnector>  Connectors = {}
*/

#include "vcc.hpp"

VCCycle::~VCCycle()
{
    delete curcon;

    for (mComponentObj::iterator iter = Comps.begin(); iter != Comps.end(); iter++)
        delete iter->second;
    Comps.clear();
}

VCCycle::VCCycle(vector<umComponent> dictComps, vector<tupConnector> vecConnectors)
{
    // 1 components
    for (auto &item : dictComps)
    {
        string classstr = any_to_string(item["classstr"]);
        string name = any_to_string(item["name"]);
        try
        { // register_type
            Comps.insert(mComponentObj::value_type(name, compinstance[classstr](item)));
        }
        catch (exception &e)
        {
            cout << e.what() << endl;
        }
    }
    // 2 connectors
    curcon = new Connector();
    for (auto &tconn : vecConnectors)
        curcon->AddConnector(tconn, Comps);

    // 3 set port addresses for all components
    for (mComponentObj::iterator iter = Comps.begin(); iter != Comps.end(); iter++)
        iter->second->setportaddress();

    Wc = 0.0;
    Qin = 0.0;
    Qout = 0.0;
    cop = 0.0;
    cop_hp = 0.0;
}

void VCCycle::component_simulator()
{
    vector<Port *> state_nodes = curcon->Nodes;

    vector<string> keys;
    for (auto &iter : Comps)
        keys.push_back(iter.first);

    bool deviceok = false;
    int CountsDev = Comps.size();
    int i = 0;

    while (deviceok == false && i <= CountsDev)
    {
        vector<string> keys_to_process = keys;
        for (auto &curdev : keys_to_process)
        {
            try
            {
                // step 2: the port state: thermal process
                Comps[curdev]->state();

                // step 3: the port state: new port's parameter pairs
                size_t j = 0;
                while (j < state_nodes.size())
                {
                    if (!state_nodes[j]->stateok)
                    {
                        state_nodes[j]->state();
                        if (state_nodes[j]->stateok)
                        {
                            state_nodes.erase(state_nodes.begin() + j);
                        }
                        else
                        {
                            j++;
                        }
                    }
                    else
                    {
                        j++;
                    }
                }

                // step 4: the port state: the energy and mass balance
                Comps[curdev]->balance();
                // Remove curdev from keys
                auto pos = find(keys.begin(), keys.end(), curdev);
                if (pos != keys.end())
                {
                    keys.erase(pos);
                }
            }
            catch (...)
            {
                // Pass the exception and continue
            }
        }
        i++;
        if (keys.size() == 0)
        {
            deviceok = true;
        }
    }

    if (keys.size() > 0)
    {
        for (auto &k : keys)
        {
            cout << k << " ";
        }
        cout << endl;
    }
}

void VCCycle::simulator()
{
    component_simulator();

    Wc = 0.0;
    Qin = 0.0;
    Qout = 0.0;

    for (auto &iter : Comps)
    {
        if (iter.second->energy == "CompressionWork")
        {
            Wc += ((Compressor *)iter.second)->Wc;
        }
        else if (iter.second->energy == "QIN")
        {
            Qin += ((Evaporator *)iter.second)->Qin;
        }
        else if (iter.second->energy == "QOUT")
        {
            Qout += ((Condenser *)iter.second)->Qout;
        }
    }
    cop = Qin / Wc;
    cop_hp = Qout / Wc;
}

void VCCycle::state()
{
    simulator();
}

void VCCycle::balance()
{
    // Already handled in simulator
}

string VCCycle::resultstr()
{
    string result;
    result = "\n --- The Cycle ---\n";
    result += "\tCompression Work(kW): " + to_string_with_precision<double>(Wc, 3) + "\n";
    result += "\tRefrigeration Capacity(kW): " + to_string_with_precision<double>(Qin, 3) + "\n";
    result += "\tCapacity(ton): " + to_string_with_precision<double>(Qin * 60.0 * (1.0 / 211.0), 3) + "\n";
    result += "\tThe heat transfer rate(kW): " + to_string_with_precision<double>(Qout, 3) + "\n";
    result += "\tThe coefficient of performance: " + to_string_with_precision<double>(cop, 3) + "\n";
    result += "\tThe coefficient of performance(heat pump): " + to_string_with_precision<double>(cop_hp, 3) + "\n";
    return result;
}

void VCCycle::outdevresultstr()
{
    for (auto &iter : Comps)
    {
        cout << iter.second->resultstring() << endl;
    }
    cout << "\n"
         << Port::title << endl;
    for (auto &item : curcon->Nodes)
    {
        cout << item->resultstring() << endl;
    }
}

void VCCycle::outresults()
{
    cout << resultstr() << endl;
    outdevresultstr();
}

File: vccrust/src/vcc.rs
//! VCCycle - Main vapor compression cycle module

use crate::common::{CompSISO, TupConnector, UMComponent, to_string_with_precision, Port};
use crate::core::Connector;
use crate::components::{Compressor, Condenser, Evaporator, ExpansionValve};
use std::collections::HashMap;

pub struct VCCycle {
    pub curcon: Connector,
    pub comps: HashMap<String, Box<dyn CompSISO>>,
    pub comp_order: Vec<String>,
    pub wc: f64,
    pub qin: f64,
    pub cop: f64,
}

impl VCCycle {
    pub fn new(dict_comps: Vec<UMComponent>, vec_connectors: Vec<TupConnector>, comp_order: Vec<String>) -> Self {
        let mut comps = HashMap::new();

        for item in dict_comps {
            let class_str = item.get("classstr")
                .and_then(|v| v.as_str())
                .unwrap_or("");
            let name = item.get("name")
                .and_then(|v| v.as_str())
                .unwrap_or("").to_string();

            match class_str {
                "Compressor" => {
                    comps.insert(name, Box::new(Compressor::new(&item)) as Box<dyn CompSISO>);
                }
                "Condenser" => {
                    comps.insert(name, Box::new(Condenser::new(&item)) as Box<dyn CompSISO>);
                }
                "Evaporator" => {
                    comps.insert(name, Box::new(Evaporator::new(&item)) as Box<dyn CompSISO>);
                }
                "ExpansionValve" => {
                    comps.insert(name, Box::new(ExpansionValve::new(&item)) as Box<dyn CompSISO>);
                }
                _ => {}
            }
        }

        let mut curcon = Connector::new();
        for tconn in vec_connectors {
            curcon.add_connector(tconn, &mut comps);
        }

        VCCycle {
            curcon,
            comps,
            comp_order,
            wc: 0.0,
            qin: 0.0,
            cop: 0.0,
        }
    }

    pub fn state(&mut self) {
        // 1 state by process - in comp order
        for name in &self.comp_order {
            if let Some(comp) = self.comps.get_mut(name) {
                comp.state();
            }
        }
        // 2 state of port/node
        for &node in &self.curcon.nodes {
            unsafe {
                (*node).state();
            }
        }
    }

    pub fn balance(&mut self) {
        // In comp order
        for name in &self.comp_order {
            if let Some(comp) = self.comps.get_mut(name) {
                comp.balance();
                if comp.energy() == "CompressorWork" {
                    if let Some(compressor) = comp.as_any().downcast_ref::<Compressor>() {
                        self.wc = compressor.wc;
                    }
                } else if comp.energy() == "RefrigerationCapacity" {
                    if let Some(evaporator) = comp.as_any().downcast_ref::<Evaporator>() {
                        self.qin = evaporator.qe;
                    }
                }
            }
        }
        self.cop = self.qin / self.wc;
    }

    pub fn resultstr(&self) -> String {
        format!(
            "\n --- The Cycle --- \
             \n\tCompression Work(kW): {} \
             \n\tRefrigeration Capacity(kW): {} \
             \n\tThe coefficient of performance: {} \n",
            to_string_with_precision(self.wc, 3),
            to_string_with_precision(self.qin, 3),
            to_string_with_precision(self.cop, 3)
        )
    }

    pub fn outdevresultstr(&self) {
        // 按C++的顺序输出：Evaporator, ExpansionValve, Condenser, Compressor
        let order = ["Evaporator", "ExpansionValve", "Condenser", "Compressor"];
        for name in order {
            if let Some(comp) = self.comps.get(name) {
                println!("{}", comp.resultstring());
            }
        }
        println!("\n{}", Port::TITLE);
        for &node in &self.curcon.nodes {
            unsafe {
                println!("{}", (*node).resultstring());
            }
        }
    }

    pub fn outresults(&self) {
        println!("{}", self.resultstr());
        self.outdevresultstr();
    }
}
